cns/restserver/util.go

```diff
@@ -160,21 +155,157 @@ func (service *HTTPRestService) saveNetworkContainerGoalState(req cns.CreateNetw
 			service.state.ContainerIDByOrchestratorContext[podInfo.PodName+podInfo.PodNamespace] = req.NetworkContainerid
 			break
 
+		case cns.KubernetesCRD:
+			// Validate and Update the SecondaryIpConfig state
+			returnCode, returnMesage := service.updateIpConfigsStateUntransacted(req, existingSecondaryIPConfigs)
+			if returnCode != 0 {
+				return returnCode, returnMesage
+			}
 		default:
 			errMsg := fmt.Sprintf("Unsupported orchestrator type: %s", service.state.OrchestratorType)
 			logger.Errorf(errMsg)
 			return UnsupportedOrchestratorType, errMsg
 		}
+
 	default:
 		errMsg := fmt.Sprintf("Unsupported network container type %s", req.NetworkContainerType)
 		logger.Errorf(errMsg)
 		return UnsupportedNetworkContainerType, errMsg
 	}
 
+	service.state.ContainerStatus[req.NetworkContainerid] =
+		containerstatus{
+			ID:                            req.NetworkContainerid,
+			VMVersion:                     req.Version,
+			CreateNetworkContainerRequest: req,
+			HostVersion:                   hostVersion}
+
 	service.saveState()
 	return 0, ""
 }
 
+// This func will compute the deltaIpConfigState which needs to be updated (Added or Deleted)
+// from the inmemory map
+// Note: Also this func is an untransacted API as the caller will take a Service lock
+func (service *HTTPRestService) updateIpConfigsStateUntransacted(req cns.CreateNetworkContainerRequest, existingSecondaryIPConfigs map[string]cns.SecondaryIPConfig) (int, string) {
+	// parse the existingSecondaryIpConfigState to find the deleted Ips
+	newIPConfigs := req.SecondaryIPConfigs
+	var tobeDeletedIpConfigs = make(map[string]cns.SecondaryIPConfig)
+
+	// Populate the ToBeDeleted list, Secondary IPs which doesnt exist in New request anymore.
+	// We will later remove them from the in-memory cache
+	if existingSecondaryIPConfigs != nil {
+		for secondaryIpId, existingIPConfig := range existingSecondaryIPConfigs {
+			_, exists := newIPConfigs[secondaryIpId]
+			if !exists {
+				// IP got removed in the updated request, add it in tobeDeletedIps
+				tobeDeletedIpConfigs[secondaryIpId] = existingIPConfig
+			}
+		}
+	}
+
+	// Validate TobeDeletedIps are ready to be deleted.
+	for ipId, _ := range tobeDeletedIpConfigs {
+		ipConfigStatus, exists := service.PodIPConfigState[ipId]
+		if exists {
+			// pod ip exists, validate if state is not allocated, else fail
+			if ipConfigStatus.State == cns.Allocated {
+				errMsg := fmt.Sprintf("Failed to delete an Allocated IP %v", ipConfigStatus)
+				return InconsistentIPConfigState, errMsg
+			}
+		}
+	}
+
+	// now actually remove the deletedIPs
+	for ipId, _ := range tobeDeletedIpConfigs {
+		service.removeToBeDeletedIpsStateUntransacted(ipId, true)
+	}
+
+	// Add the newIpConfigs, ignore if ip state is already in the map
+	service.addIPConfigStateUntransacted(req.NetworkContainerid, newIPConfigs)
+
+	return 0, ""
+}
+
+//AddIPConfigsToState takes a lock on the service object, and will add an array of ipconfigs to the CNS Service.
```

Comment:
The func documentation seems misleading. It has been renamed and is no longer taking locks.